### PR TITLE
Maintain compatibility with Rust 1.25

### PR DESCRIPTION
**Pull request by azriel91**
Issue #135

Not sure if you want to also include a CI config for Rust 1.18

### CODE REVIEW

**Comment by azriel91** [comment]
Implementation is incorrect for detecting rust features, shall find another way.

**Comment by azriel91** [comment]
Ah, 1.18.0 built on windows, but not linux. I'm off for tonight, but feel free to add to the branch if it's urgent.

**Comment by alexcrichton** [comment]
This is unfortunately a bit too much code duplication for me to be comfortable maintaining purely for the reason of supporting older compilers. Is there a compelling reason to support 1.18.0?

**Comment by azriel91** [comment]
Hm, besides "make the `failure` PR pass", I think that's better answered by @burntsushi

**Comment by BurntSushi** [comment]
I've honestly lost a lot of my own motivation for tracking older Rust releases. It is too much work, and in particular, is completely intractable without community consensus on what to do. I think we should at the very minimum put a specific Rust version in our CI config so that folks can at least know which version is supported though.

**Comment by RalfJung** [comment]
`error_len` is stable since Rust 1.20.  Assuming that this removes most of the need for code duplication, would it be okay for failure to bump its minimal version to 1.20?

**Comment by alexcrichton** [comment]
Heh to add to what @BurntSushi says I've also lost motivation to even list the minimum Rust version in CI, all it ends up doing for me is making it annoying to land PRs that bump it (as you always forget about it) and also confusing contributors who aren't typically aware that it can be bumped. 

It looks like the main duplication here is happening at the rust 1.25.0 boundary, so it seems fine to me to land whatever's necessary to support 1.25.0+ as it likely doesn't involve much duplication

**Comment by RalfJung** [comment]
Oh dang I missed the `repr(align)` thing and that it requires Rust 1.25.

**Comment by azriel91** [comment]
Okay, so with 1.25.0 as the minimum, things are much easier!
Actually many times when Rust releases a new stable version, I think "wow, this should be the minimum stable" 😅.

File: src/symbolize/dbghelp.rs [hunk 1]
```diff
@@ -10,8 +10,12 @@
 
 #![allow(bad_style)]
 
+#[cfg(feature = "std")]
+use std::char;
+
 use core::mem;
 use core::slice;
+#[cfg(not(feature = "std"))]
```
Comment (alexcrichton) [inline src/symbolize/dbghelp.rs]: Hm what's going on here? Was this something we moved from std to core in the last few versions?

**Comment by alexcrichton** [comment]
Looks great to me! Just one thing I'm curious about, and it also looks like CI may be failing for 1.25.0? (I don't think std has `c_void` by that point)

File: src/symbolize/dbghelp.rs (hunk 1 above)
Comment (azriel91) [reply src/symbolize/dbghelp.rs]: Yeap, based on issue https://github.com/rust-lang/rust/issues/49319 and the corresponding PR: https://github.com/rust-lang/rust/pull/49698 which happened on 2018-04-12, which is *just* after Rust 1.26 was [released](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1260-2018-05-10).

**Comment by azriel91** [comment]
Yeap, took longer than I'd've liked to figure out that `no_std` support isn't available until Rust 1.30. I've updated the CI config to test the features with `std`, though it duplicates nearly all of the main script. Yaml doesn't have a standard merge syntax (probably for good reason).

**Comment by azriel91** [comment]
So, the `"gimli-symbolize"` feature transitively uses `parity-wasm` (backtrace -> addr2line 0.7 -> object 0.9 -> parity-wasm 0.31.3).

`parity-wasm` makes use of [auto dereferencing match patterns](https://github.com/rust-lang/rust/pull/49394) [introduced in Rust 1.26](https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-1260-2018-05-10).

Road from here, things that I'd consider unfeasible (maintenance cost, sanity):

* Chase down every crate and "make it backward compatible"
* Track down Rust compatibility for every dependency and introduce Rust version checks for each.

I've removed it from the 1.25 CI tests.

**Comment by alexcrichton** [comment]
Thanks so much!

**Pull request merged**